### PR TITLE
feat: vertical split files when L0s are highly overlapped

### PR DESCRIPTION
**Pull request by Joe-Blount**
This PR adds a "vertical split" (splitting at a common set of times a set of files that may total >>>`max_compact_size`) in cases where the L0s are heavily overlapped.  Without this, the compaction can get stuck in a pattern where a set of 2-3 `max_compact_size` files get recompacted and split at the same times as previously, so the compaction doesn't accomplish anything.  Once in that state, those unproductive compaction will loop until it times out.  This PR avoids that scenario by recognizing the bigger picture and performing a split across more files than we otherwise would.  By splitting a larger set of L0s into a smaller time range, we're put in position the next round of compactions to be much more productive.

Currently this fix only helps scenarios where we're over the `max_num_files_per_plan`.  Addressing the situation for <`max_num_files_per_plan` may require changing the classification criteria or moving this logic, but one step at a time.

helps: https://github.com/influxdata/idpe/issues/17292
helps: https://github.com/influxdata/idpe/issues/17283

- [x] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [x] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

### CODE REVIEW

File: compactor2/src/components/split_or_compact/start_level_files_to_split.rs [hunk 1]
```diff
@@ -7,12 +7,161 @@ use crate::{
...
+                    });
+                } else {
+                    files_not_to_split.push((*file).clone());
+                }
+            }
+
+            break;
+        }
+
+        overlaps = vec![];
```
[the remark this answers is not in the document]
Comment (carols10cents) [reply compactor2/src/components/split_or_compact/start_level_files_to_split.rs:139]: Yes, we could do something like `files_to_split.files().iter().any(|f| f==file)`, but I'm wondering if it's really needed or not...

Right now, it looks like you're considering each `file` (line 86, `for file in &files`) and comparing each to all other files (line 92, `overlaps = files.iter().filter(|f| file.overlaps(f)).collect();`), which both take immutable references, then `file` has to be cloned when you put it into either `files_to_split` or `files_not_to_split`. 

oops meeting is back on, I'll come back and finish this thought later but I want to save this

File: compactor2/src/components/split_or_compact/start_level_files_to_split.rs [hunk 2]
```diff
@@ -7,12 +7,161 @@ use crate::{
...
+    // The files in `files` all overlap, but they might overlap in a chain (file1 overlaps file1, which overlaps file3...)
+    // This algorithm is not concerned with the chain of overlaps that keeps pulling in more files.  Instead, for each file,
+    // we'll look at what overlaps just it.  If that's over max_compact_size, we'll split them.
+    for file in &files {
+        if file.compaction_level != start_level {
+            continue;
+        }
+
+        // Get the set of overlaping start_level files that includes file
+        overlaps = files.iter().filter(|f| file.overlaps(f)).collect();
```
Comment (NGA-TRAN) [inline compactor2/src/components/split_or_compact/start_level_files_to_split.rs:92]: Since `files` also include the `file`, I think this overlaps will include all files, right? Which means not what you want? Maybe you have to check `f != file`, too
```suggestion
        overlaps = files.iter().filter(|f|  f!file &&   file.overlaps(f)).collect();
```

I have not verified if it is compiled

File: compactor2/src/components/split_or_compact/start_level_files_to_split.rs [hunk 3]
```diff
@@ -7,12 +7,161 @@ use crate::{
...
+
+        let min = overlaps.iter().map(|f| f.min_time.get()).min().unwrap();
+        let max = overlaps.iter().map(|f| f.max_time.get()).max().unwrap();
```
Comment (NGA-TRAN) [inline compactor2/src/components/split_or_compact/start_level_files_to_split.rs:95]: If overlaps is empty, we will hit panic. We avoid using `unwrap` in non-test code unless we know for sure it is not empty

File: compactor2/src/components/split_or_compact/start_level_files_to_split.rs [hunk 4]
```diff
@@ -7,12 +7,161 @@ use crate::{
...
+            // get min time, max time, split it into (1 + cap/max) pieces
+            let delta = (max - min) / (splits + 1) as i64;
+            if delta == 0 {
+                // TODO: this is very small time range, but if the delta is non-zero we could still split it.
+                overlaps = vec![];
+                continue;
+            }
+            let split_times: Vec<i64> = (1..splits + 1).map(|i| min + (i as i64 * delta)).collect();
+
+            assert!(!split_times.is_empty());
```
Comment (NGA-TRAN) [inline compactor2/src/components/split_or_compact/start_level_files_to_split.rs:116]: I honestly do not get the how this split computation/reasoning works but if it works, it is good enough. 
I suggest to move lines 100 to 116 to a function. If possible, have some examples of input and output to explain how they are split. I think this function itself deserves some unit tests.

File: compactor2/src/components/split_or_compact/start_level_files_to_split.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (NGA-TRAN) [reply compactor2/src/components/split_or_compact/start_level_files_to_split.rs:139]: @carols10cents : can help review the full PR. I have concerns about the overlap. See comments above.
I think what Joe wants is  `into_iter` to get the file out of the list

File: compactor2/tests/layouts/large_files.rs [hunk 5]
```diff
@@ -626,65 +626,65 @@ async fn two_large_files_total_over_max_compact_size_start_l0() {
...
     - "L0, all files 150mb                                                                                                "
     - "L0.1[0,1000] 10ns        |------------------------------------------L0.1------------------------------------------|"
     - "**** 2 Output Files (parquet_file_id not yet assigned), 150mb total:"
-    - "L0                                                                                                                 "
-    - "L0.?[0,667] 10ns 100.05mb|---------------------------L0.?---------------------------|                              "
-    - "L0.?[668,1000] 10ns 49.95mb                                                            |-----------L0.?------------| "
-    - "**** Simulation run 1, type=split(split_times=[667]). 1 Input Files, 150mb total:"
+    - "L0, all files 75mb                                                                                                 "
+    - "L0.?[0,500] 10ns         |-------------------L0.?--------------------|                                             "
+    - "L0.?[501,1000] 10ns                                                   |-------------------L0.?-------------------| "
```
Comment (NGA-TRAN) [inline compactor2/tests/layouts/large_files.rs:635]: Is this the reason now that we do not split according to max_desired_file_byte but your new formula?

File: compactor2/tests/layouts/many_files.rs [hunk 6]
```diff
@@ -3771,600 +3772,4542 @@ async fn random_backfill_empty_partition() {
...
+    - "L0.?[50,356] 1.01us 5.28mb|--------------------L0.?---------------------|                                           "
+    - "L0.?[357,629] 1.01us 4.72mb                                               |------------------L0.?------------------| "
+    - "**** Simulation run 39, type=split(split_times=[356, 670]). 1 Input Files, 10mb total:"
+    - "L0, all files 10mb                                                                                                 "
+    - "L0.11[173,950] 1.01us    |-----------------------------------------L0.11------------------------------------------|"
+    - "**** 3 Output Files (parquet_file_id not yet assigned), 10mb total:"
+    - "L0                                                                                                                 "
+    - "L0.?[173,356] 1.01us 2.36mb|-------L0.?--------|                                                                     "
+    - "L0.?[357,670] 1.01us 4.03mb                     |---------------L0.?---------------|                                 "
+    - "L0.?[671,950] 1.01us 3.62mb                                                         |-------------L0.?-------------| "
```
Comment (NGA-TRAN) [inline compactor2/tests/layouts/many_files.rs:4084]: Nice to vertically split all files at 356 and 6770

File: compactor2/tests/layouts/many_files.rs [hunk 7]
```diff
@@ -3771,600 +3772,4542 @@ async fn random_backfill_empty_partition() {
...
+    - "L0.109[76,356] 1.03us 3.27mb   |---------L0.109---------|                                                             "
+    - "L0.108[671,986] 1.03us 3.36mb                                                           |-----------L0.108-----------| "
+    - "L0.107[357,670] 1.03us 3.32mb                              |----------L0.107-----------|                               "
+    - "L0.106[42,356] 1.03us 3.33mb|----------L0.106-----------|                                                             "
+    - "**** 3 Output Files (parquet_file_id not yet assigned), 300mb total:"
+    - "L1                                                                                                                 "
+    - "L1.?[42,357] 1.03us 100.11mb|------------L1.?------------|                                                            "
+    - "L1.?[358,672] 1.03us 99.79mb                              |-----------L1.?------------|                               "
+    - "L1.?[673,986] 1.03us 100.11mb                                                            |-----------L1.?------------| "
```
[the remark this answers is not in the document]
Comment (NGA-TRAN) [reply compactor2/tests/layouts/many_files.rs:4255]: I think maybe a ticket for the future is instead of splitting the results at 357 and 672, we will keep the same splits at 356 and 670. I know this is not trivial but if we open a  ticket and point to this test, we will  have a context for optimization.

If we optimize this, we will avoid runs 51 to 85

**Review by NGA-TRAN — COMMENTED** [review]
Very nice split function @Joe-Blount 🎉 

I only have a few rust code concerns that may lead to panic and extra overlapped files. I have asked @carols10cents to review.

git does let me review `knobs.rs` because of too much changes. However, the added lines < deleted lines which is a good sign. Have you verified them carefully?

File: compactor2/src/components/split_or_compact/start_level_files_to_split.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (carols10cents) [reply compactor2/src/components/split_or_compact/start_level_files_to_split.rs:139]: Yes, I think Nga and I are getting to the same thing: is it possible to *move* each `file` as they're processed, and then only check overlaps with the remaining, unprocessed files? I haven't yet thought through if this will result in the same algorithm as what you have here, and there may be borrow-checker implications here that might influence how it ends up looking, but does that make sense and what do you think?

**Comment by carols10cents** [comment]
Yeah, I think having tests for smaller parts of this would help a lot. And sorry for causing conflicts-- if you'd like me to try a rebase, I'd be happy to!

File: compactor2/src/components/split_or_compact/start_level_files_to_split.rs (hunk 3 above)
Comment (Joe-Blount) [reply compactor2/src/components/split_or_compact/start_level_files_to_split.rs:95]: Currently since `overlaps` includes `file` we're checking for overlaps, it cannot be empty.  But I can add a test case where `file` is the only file in `overlap` (it won't panic in the unwrap, but that's still an interesting case to test...)

File: compactor2/src/components/split_or_compact/start_level_files_to_split.rs (hunk 2 above)
Comment (NGA-TRAN) [reply compactor2/src/components/split_or_compact/start_level_files_to_split.rs:92]: If so so, `overlaps` is actually the full set of `files` with start_level, right?
If that is what you need, you only need iterator with a filter to keep only start_level files

File: compactor2/src/components/split_or_compact/start_level_files_to_split.rs (hunk 4 above)
Comment (Joe-Blount) [reply compactor2/src/components/split_or_compact/start_level_files_to_split.rs:116]: I'll push a commit that moves those lines to a new function and adding specific test cases.  Also adding more comments to explain the though process.
That clarity lead me to tweak the math slightly to get better boundary condition handling (which unfortunately churns some details in some of the insta checks).

File: compactor2/src/components/split_or_compact/start_level_files_to_split.rs (hunk 2 above)
Comment (Joe-Blount) [reply compactor2/src/components/split_or_compact/start_level_files_to_split.rs:92]: > If so so, `overlaps` is actually the full set of `files` with start_level, right?

no.
`files` is the full chain of overlapping files.

For example, for this:
```
        - "L0.2[650,750] 180s                                                    |------L0.2------|                           "
        - "L0.1[450,620] 120s                |------------L0.1------------|                                                   "
        - "L0.3[800,900] 300s                                                                               |------L0.3------|"
        - "L1, all files 100b                                                                                                 "
        - "L1.13[600,700] 60s                                           |-----L1.13------|                                    "
        - "L1.12[400,500] 60s       |-----L1.12------|                                                                        "
```
`files` could be L1.12, L0.1, L1.13, L0.2 I call that a "chain" because L1.12 overlaps L0.1, which overlaps L1,13, which overlaps L0.2.   L1.12 does not overlap L0.2, but they're in `files` together because a chain of overlaps that doesn't break until L0.3, which doesn't overlap anything.

This algorithm selects narrower sets of files, not the whole chain.
So if the for loop starts with L0.2, `overlaps` would contain only L0.2 and L1.13 because L1.13 is the only thing that overlaps L0.2.  

Think of the logic as:
If following the chain pulls in too much and we can't make good decisions on it, instead of following the chain we should pick an L0 and focus just on what we need to for it.

File: compactor2/tests/layouts/many_files.rs [hunk 8]
```diff
@@ -3771,600 +3772,4542 @@ async fn random_backfill_empty_partition() {
...
+    - "  Upgrading 2 files level to CompactionLevel::L2: L1.560, L1.568"
+    - "  Creating 3 files"
+    - "**** Final Output Files "
+    - "L2                                                                                                                 "
+    - "L2.560[42,268] 1.05us 100mb|------L2.560-------|                                                                     "
+    - "L2.568[269,445] 1.05us 100.08mb                     |----L2.568----|                                                     "
+    - "L2.570[446,627] 1.05us 100.53mb                                      |----L2.570-----|                                   "
+    - "L2.571[628,808] 1.05us 99.97mb                                                       |----L2.571-----|                  "
+    - "L2.572[809,986] 1.05us 99.42mb                                                                         |----L2.572----| "
+    "###
```
[the remark this answers is not in the document]
Comment (Joe-Blount) [reply compactor2/tests/layouts/many_files.rs:5905]: yes, but most of the change was from changing changing hte test case by setting `with_max_num_files_per_plan` higher.

File: compactor2/tests/layouts/large_files.rs (hunk 5 above)
Comment (Joe-Blount) [reply compactor2/tests/layouts/large_files.rs:635]: yes.  The key criteria for this situation is: 
- the (2) overlapping files total more than the max compact size because each is slightly over half the max compact size.  This triggers the  new logic.
- since the new logic is triggered, where overlapping L0s are more than the max compact size, it's trying to split in a way that minimizes the risk of the split result still exceeding the max compact size.  Since its doing one split, it splits in the middle of the time range.  

In real data the data probably isn't dispersed linearly across the time range, so splitting right in the middle is the least likely to require another split.

File: compactor2/src/components/split_or_compact/start_level_files_to_split.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Joe-Blount) [reply compactor2/src/components/split_or_compact/start_level_files_to_split.rs:130]: I found something saying I can modify a vector while iterating it, if its immediately followed by a break statement.
That would have worked, except that `files` is also borrowed as immutable, so that leads to mutable borrow after immutable borrow errors.

File: compactor2/src/components/split_or_compact/start_level_files_to_split.rs (hunk 2 above)
Comment (NGA-TRAN) [reply compactor2/src/components/split_or_compact/start_level_files_to_split.rs:92]: The explanation makes sense. Thanks @Joe-Blount 

I have  different question/comment. Let us look at the example above and which is also the test I commented above. Why in that test, L0.1 (and its overlapped L1.12) is not selected but instead L0.2 (and its overlapped L1.13) is selected. I guess because L0.2 happens to be in front of L0.1 in the `files`. I think since we pick files with smallest `max_l0_created_at` to compact first, splitting those files first will provide higher chance to reduce the number of rounds to compact and split. 

What do you think to order `files` in the order of `max_l0_created_at` first and then run the loop to select what to split?

File: compactor2/src/components/split_or_compact/start_level_files_to_split.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (NGA-TRAN) [reply compactor2/src/components/split_or_compact/start_level_files_to_split.rs:130]: I have reread the algorithm, it looks fine with your intention. It looks good as as except to order the `files` before the loop in my comment above

**Review by NGA-TRAN — COMMENTED** [review]
Looks really nice @Joe-Blount . I only have 2 comments that I think worth your time considering before merging this to main. We can have a quick chat as needed, too.

File: compactor2/src/components/split_or_compact/split_compact.rs [hunk 9]
```diff
@@ -341,14 +356,15 @@ mod tests {
...
+        - "L1, all files 100b                                                                                                 "
+        - "L1.13[600,700] 60s       |--------------------------L1.13---------------------------|                              "
         - "files to keep:"
         - "L0, all files 100b                                                                                                 "
-        - "L0.2[650,750] 180s                                                    |------L0.2------|                           "
+        - "L0.1[450,620] 120s                |------------L0.1------------|                                                   "
         - "L0.3[800,900] 300s                                                                               |------L0.3------|"
         - "L1, all files 100b                                                                                                 "
         - "L1.12[400,500] 60s       |-----L1.12------|                                                                        "
-        - "L1.13[600,700] 60s                                           |-----L1.13------|                                    "
```
[the remark this answers is not in the document]
Comment (Joe-Blount) [reply compactor2/src/components/split_or_compact/split_compact.rs:351]: L0.1 and L1.2 were not picked because I wasn't sorting the list before the loop, as you mentioned in other comments.
I'll restore this test case to its original intent, and add a new one :+1:

File: compactor2/src/components/split_or_compact/start_level_files_to_split.rs (hunk 2 above)
Comment (Joe-Blount) [reply compactor2/src/components/split_or_compact/start_level_files_to_split.rs:92]: Great point! yes, I should sort the files before the loop.

File: compactor2/src/components/split_or_compact/split_compact.rs [hunk 10]
```diff
@@ -356,15 +404,14 @@ mod tests {
         ---
         - "files to compact or split:"
         - "L0, all files 100b                                                                                                 "
-        - "L0.2[650,750] 180s                                     |---------------------------L0.2---------------------------|"
+        - "L0.1[450,620] 120s                      |----------------------L0.1-----------------------|                        "
         - "L1, all files 100b                                                                                                 "
-        - "L1.13[600,700] 60s       |--------------------------L1.13---------------------------|                              "
+        - "L1.13[600,700] 60s                                                                   |-----------L1.13------------|"
+        - "L1.12[400,500] 60s       |-----------L1.12------------|                                                            "
```
Comment (NGA-TRAN) [inline compactor2/src/components/split_or_compact/split_compact.rs:410]: Nice 🎉

**Review by NGA-TRAN — APPROVED** [review]
LGTM
